### ISSUE TITLE
Make Quick Open use the new Preview editor to display results (like sublime's find anything)

### ISSUE THREAD

**Issue by drew-wallace** [Microsoft/vscode#8939]
- VSCode Version: 1.3.0
- OS Version: Windows 10 64bit & Mac OS X 10.11.3

Steps to Reproduce:
1. Open the Quick Open box
2. Type a file name that returns several results
3. A new Preview editor pops up and receives focus with the first result loaded into it
4. Scroll through the results in the Quick Open box and the Preview editor changes to the file that's highlighted
5. Selecting a result will fully open the file. Canceling can close the Preview editor or leave it open. I'd prefer if it closed the preview.

**Comment by bgashler1** [comment]
I myself have been wanting this very same thing.  I tend to open several files up front, and then I have to double-click each one.

I think that opening something via Quick Open as a fully-opened file makes sense in this case, because if you had the opportunity to see a preview and still committed via enter or a click to open it then you probably intend to keep it around for a while.

**Comment by drew-wallace** [comment]
Just to be clear, the point of this is to see the contents of a file as I scroll through the results of Quick Open. As far as permanently opening a file from quick open, that would be good too.

**Comment by stevencl** [comment]
Previewing files as you move up and down the list of results makes a lot of sense. For files that are currently open we would bring them to the front.

**Comment by smuuf** [comment]
I'm glad I found this issue *(under this not-so-much clear title)*, so I don't have to create it myself.

I just got annoyed by this behaviour of closing files - I evidently wanted to have opened - for the 451th time in the last two weeks.

*Please, let's make it into the next release.* :)

**Comment by wprater** [comment]
Ive spent some time working through this.  Have a good understanding of the organization of the editorPart, groups, stack model, etc.  Pushing up a PR, but I'd really like to fix some UX issues below.

> The preview editor will remain and whatever you last were looking at in Quick Open will also remain inside that preview editor—exception: if you were on a preview editor when you opened Quick Open and previewed some items but never selected any (e.g. escaped out of it) then we will restore whatever you previously had open in that preview tab, since that would indicate that the preview tab is likely something you want to keep and not replace.

I think a better UX would be to close a preview editor if there were none open.  Hitting escape should remove all side effects e.g. opening a new preview editor.

> Upon hitting Enter, the open editor will be kept open (i.e. no longer preview).

Would we prefer to honor the user preference `workbench.editor.enablePreviewFromQuickOpen` (this is how Ive implemented it)

**Comment by wprater** [comment]
Another UX issue that I discovered while previewing in a current `previewEditor`: the scroll positions are lost.  

We could keep track of a selection/cursor when opening with `TextEditorOptions`, however, this feels kind of kludgy to keep track of this just if we are swapping the contents of a preview editor.  An easier approach could be to render a tab-less editor that would be show while previewing files that are not find in the current editor groups.  

Thoughts?

**Comment by drew-wallace** [comment]
Sublime preserves the cursor position of each file, and when you preview that file, it shows the cursor in that position. If you haven't visited that file before, it defaults to the top of the file.

**Comment by wprater** [comment]
> Sublime preserves the cursor position of each file, and when you preview that file, it shows the cursor in that position. If you haven't visited that file before, it defaults to the top of the file.

actually, they use a brand new editor buffer when displaying a file that has no editor associated with it.  otherwise, you're effectively just swapping the viewable editors (VS Code also will preserve line and viewport position in this respect).  this is what Im doing now in my PR.  

Im lobbying for a new buffer with no tab, so we can have a similar experience to what ST has.

**Comment by drew-wallace** [comment]
Also, Sublime will put focus on an open tab if what you have selected in Quick View is already open.

**Comment by wprater** [comment]
yeah, see my PR.  these behaviors are implemented.

**Comment by bpasero** [comment]
@wprater yes, the preview editor has to go away when I cancel my quick open session without picking an item. I cannot see how the scroll positions can get lost unless you actually provide options with line/col information. We always persist the view state of the editor and restore it on open unless you provide your own view state.

**Comment by wprater** [comment]
> We always persist the view state of the editor and restore it on open unless you provide your own view state.

The issue is that I need to render another temporary editor and display it as an `eager` preview.  Its not really a `preview` (or `pinned`) editor until one selects it.

Do you know of a method in which I could create such an editor?  I can keep looking around, but if there are some mechanics already, I'd like to utilize them.

**Comment by bpasero** [comment]
@wprater I do not understand why you would want to open a temporary editor...

**Comment by wprater** [comment]
> @wprater I do not understand why you would want to open a temporary editor...

Might make it easier around the logic of maintaining when a new preview editor was opened or not.  As well as the issue Im having when previewing in the preview editor, where the new files seem to be opened at the final line, vs. the top of file (they were not previously opened.)

If you did not have a preview editor open and the eager preview display opened an editor, then when cancelling, Im not certain how rest the editor group's state of active/preview editors.

**Comment by bpasero** [comment]
Whatever we end up doing, you cannot open a "preview editor", it can only be the file imho. I think in order to restore the previous state you probably have to capture it and restore it. 

As for opening on first line, you can pass options for the open to have it open at line/col 1. it is interesting if that is the desired behaviour though.

**Comment by wprater** [comment]
Maybe my nomenclature is a bit off.  I believe the current PR follows suit to the UX meetings notes.  However, I would prefer we not allow any side effect when cancelling out of a quick open widget.  If you agree, I need to determine if a new preview editor was opened when not one presently.  If there was not one open, I need to be sure to close the preview editor that was created to preview the focused items in the tree.  I'll look into that again tomorrow.

I don't desire to have them open on the first line/col, unless the file was not already present in an editor buffer.  If I continually swap the preview editors, I was seeing many of my files loaded into the preview on the last line.  The other files editors that are selected seem to work great!

**Comment by bpasero** [comment]
Ok I think I was confused by terms here. We agree on this behaviour:
- editors open as files when you navigate through the list
- they open as preview editors (not pinned) no matter what the user has configured to prevent accumulation of many editors
- they open on whatever line/col they used to be
- if there was no preview editor opened, the preview editor closes upon cancel
- if there was a preview editor, it gets restored upon cancel

**Comment by wprater** [comment]
thanks for outlining these.  

I have these all complete in the PR with exception of comments below:

> if there was no preview editor opened, the preview editor closes upon cancel

need to determine best way to achieve this functionality.

also, can you edit your list to include a not about history?  I want to be sure Im not including history events.  there are two types, from what I can tell: `handleEditorEventInStack` and `handleEditorEventInHistory`.  Do we need stop inclusion in both, or just the stacks?

**Comment by bpasero** [comment]
@wprater [IEditorGroupService.getStacksModel()](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/services/group/common/groupService.ts#L86) gives you complete access about what is going on in the editor stacks world (1 stack contains 3 groups side by side of N editor tabs). I think the best is to just store the current preview editor before doing the magic and then restore either the previous one on cancel or close the opened one if there was no preview editor before.

Yes you are quite right that these are the two methods that need to ignore the preview editors. The former is a stack of editors that you can go back and forward via the Go menu and the latter is the editor history that is persisted across restarts and accessible from Cmd+P. As you can see there is already a boolean `blockStackChanges` for the stack and we need to have a new method that flips another boolean for the duration of the previewing that also blocks changes to history imho.

**Comment by bgashler1** [comment]
Updated design notes above.  Thanks :)

**Comment by bsvipas** [comment]
@bpasero @roblourens Right now if we use CMD + P with right arrow key it will open file, but without preview mode. So basically if we improve right arrow key in file navigation to open file in preview mode this feature will be done. Of course it will be a little bit different from ST 3, because you see file content simultaneously while navigating.

**Comment by aaronmw** [comment]
What's happening with this feature? It looks like it's been discussed since 2016 but I'm not seeing it in the latest VS Code, even with `workbench.editor.enablePreview` and `workbench.editor.enablePreviewFromQuickOpen` set to `true`

**Comment by bsvipas** [comment]
I'm interested too because this feature in Sublime is OP, we definitely need this in VS Code!!!

**Comment by stevencl** [comment]
Unfortunately it seems like there are some technical issues blocking this. See this comment here: https://github.com/Microsoft/vscode/issues/12892#issuecomment-286149590

**Comment by redbar0n** [comment]
@stevencl @bpasero The issues mentioned in 
[#12892 (comment)](https://github.com/Microsoft/vscode/issues/12892#issuecomment-286149590)  should not be a blocker. It could simply just work like preview mode currently does when selecting any new file from the file explorer. It will open a preview tab, and if selecting another file, then the preview tab name and content changes. It's not necessary to augment the core workbench UI with a whole new way of viewing files without any visible tab.

Wouldn't this be easy enough to implement if reusing the preview mode from the file explorer?

**Comment by ZYinMD** [comment]
I submitted #60822 to request a feature to immediately switch tabs when pressing Ctrl-Tab (as opposed to switching when Ctrl is released), but it was closed for being a duplicate of this one. 

I don't see why it's a duplicate. I mean it's sort of related, but that one was much easier to implement. Maybe they're planning something big?

**Comment by guisehn** [comment]
Any news? Trying to switch from Sublime to VS Code and I miss this feature a lot.

**Comment by yansern** [comment]
I need this so badly!

**Comment by mmcintyre123** [comment]
There is a very nice extension that has begun to implement this functionality.  It could use some improvement if anyone was the time to contribute (and if the author is accepting contributions of course).

https://marketplace.visualstudio.com/items?itemName=Cmacu.gotoanything
https://github.com/Cmacu/gotoanything

**Comment by Cmacu** [comment]
@mmcintyre123 Thanks.
I will look into ways to improve the performance and add other features this weekend.
Contributions are welcome.

**Comment by ilyakamens** [comment]
I'd also very much like this feature!

**Comment by vinceprofeta** [comment]
What is the issue here? This is a show stopper for us, any plan to release this functionality?

**Comment by indigoram89** [comment]
I surprised that vscode does not support it....

**Comment by aPinix** [comment]
Yes, Sublime Text has it and it's awesome. Any news on this?

**Comment by swist666** [comment]
> Yes, Sublime Text has it and it's awesome. Any news on this?

+1

**Comment by Dylan-Kerler** [comment]
Is there an open PR for this? I would like to contribute if possible

**Comment by r-moore** [comment]
> Yes, Sublime Text has it and it's awesome. Any news on this?

+1

**Comment by rogiervandenberg** [comment]
For now, I'd just go for https://marketplace.visualstudio.com/items?itemName=Cmacu.gotoanything Does exactly this :)

**Comment by remainstheday** [comment]
![68747470733a2f2f692e696d6775722e636f6d2f5454546a6136742e676966](https://user-images.githubusercontent.com/4063510/155036251-a91c8fb3-f971-4781-b4af-ec07dd910f58.gif)
What's the status??? This feature was a major performance enhancer for me while playing with VIM.

**Comment by trebor86** [comment]
Things I miss in quick open after switching to VSC compared to Sublime Text:
- file preview
- file path in a separate row

**Comment by trebor86** [comment]
> For now, I'd just go for https://marketplace.visualstudio.com/items?itemName=Cmacu.gotoanything Does exactly this :)

Not exactly, it messes up searching with multiple words, e.g. if I have a file path c:/fruits/bananas/File.js searching with "bananas file" won't return results. Without the plugin it does.

**Comment by jespertheend** [comment]
It also doesn't show file icons in front of the results. I've submitted https://github.com/Cmacu/gotoanything/issues/8 and https://github.com/Cmacu/gotoanything/issues/9

**Comment by Cmacu** [comment]
> Not exactly, it messes up searching with multiple words, e.g. if I have a file path c:/fruits/bananas/File.js searching with "bananas file" doesn't return results. Without the plugin it does.

> It also doesn't show file icons in front of the results. I've submitted [Cmacu/gotoanything#8](https://github.com/Cmacu/gotoanything/issues/8) and [Cmacu/gotoanything#9](https://github.com/Cmacu/gotoanything/issues/9)

Got it. I can resolve both of these over the weekend.

**Comment by TylerLeonhardt** [comment]
I feel like there's prior art here in the suggest widget:
![Image](https://user-images.githubusercontent.com/2644648/217641929-ca57567d-4ec8-453b-b39b-e8daae407007.png)

**Comment by dpidan** [comment]
> I feel like there's prior art here in the suggest widget: 

@TylerLeonhardt, Which dialog is that? It looks like a different dialog than the "quick open, go to file" (Crtl/CMD-p) which the OP seems to indicate in step 2:

> 2. Type a file name that returns several results

**Comment by jespertheend** [comment]
@dpidan I think the screenshot is referring to one of VSCode apis.
Basically if you press the right arrow with one of the files selected from quick open, it opens the file without closing the quick open dialog.
The only issue is that those files are opened permanently, so pressing the right arrow key subsequently on different files will just open a whole bunch of tabs, rather than replacing the previous editor.

**Comment by TylerLeonhardt** [comment]
@dpidan it's Intellisense in an editor... that has items, with additional context, and is keyboard driven.

**Comment by popadicbranislav** [comment]
While we wait for this feature, this extension [FindItFaster](https://marketplace.visualstudio.com/items?itemName=TomRijndorp.find-it-faster) does a pretty good job. It would be great if we get this built-in.

**Comment by codercodingthecode** [comment]
7 years later and counting ⏳ 😄

**Comment by Tantol** [comment]
> While we wait for this feature, this extension [FindItFaster](https://marketplace.visualstudio.com/items?itemName=TomRijndorp.find-it-faster) does a pretty good job. It would be great if we get this built-in.

It will be amazing to have fzf with that preview build in to VsCode.

**Comment by lakshits11** [comment]
Dear Diary,
Its been December 2023. This feature is not implemented in VS Code. I will share more updates later.

:(

**Comment by martindrzka** [comment]
I'd pay for that feature. I really miss it.

**Comment by h0jeZvgoxFepBQ2C** [comment]
This feature is missing so much... PLEEEEESEEE

**Comment by joshmu** [comment]
I made a little extension 🫧 [Periscope](https://marketplace.visualstudio.com/items?itemName=JoshMu.periscope) last year to give me an experience as close as possible whilst constrained by the existing vscode UI primitives.

It's evolved a bit to lean in to ripgrep but its remained my daily driver for grepping across the workspace.

Might be useful to some of you! :v:

**Comment by macintacos** [comment]
@joshmu this kinda looks like the built-in "Quick Search" feature that VSCode has been iterating on for the past few releases, no? https://code.visualstudio.com/updates/v1_88#_quick-search-improvements

**Comment by jbrodriguez** [comment]
> @joshmu this kinda looks like the built-in "Quick Search" feature that VSCode has been iterating on for the past few releases, no? https://code.visualstudio.com/updates/v1_88#_quick-search-improvements

however, there's no quick preview, i would guess it's because of the limitations with the vscode ui primitives, as @joshmu mentioned

that telescope.nvim ui is wicked, i'm seriously considering nvim

**Comment by macintacos** [comment]
@jbrodriguez not sure what you mean. As you arrow-through the results using VSCode's Quick Search, the currently-focused editor shows you a preview of that result in the file where it exists. Seems the same as the video in the extension I was replying to.

**Comment by jbrodriguez** [comment]
you're right, yes, not as straightforward, but does the job

**Comment by joshmu** [comment]
@macintacos the **Quick Search (experimental)** is similar to **Periscope** in regards to file previews as you navigate across results.  The biggest difference would be that Periscope provides access to more of **ripgrep**.  eg: pattern matching, file type & sub folder filtering, etc.  You can personalise this to your needs and create abbreviations for your common searches.  

If there is no interest in those type of advanced grep features then Quick Search would be sufficient

**Comment by h0jeZvgoxFepBQ2C** [comment]
@macintacos but the quicksearch feature which you mention, does not work when you go to "open file" search dialogue or? Just for full text seach, not for file search.

**Comment by macintacos** [comment]
@h0jeZvgoxFepBQ2C right. I wasn't saying that that feature in VSCode takes care of this issue. I was just saying that that feature is very similar to the Periscope extension that was linked to earlier in the thread.

I don't think that the issue described in the original description is "solved", by any means.

**Comment by musjj** [comment]
Would love to have this feature in the editor picker (when you press <kbd>Ctrl+Tab</kbd>) too.

**Comment by LucianCrainic** [comment]
Any news from the dev team on this ? A built in telescope in vscode would make the programming experience soo much better...

**Comment by popielasty** [comment]
Bump, still waiting ;)

**Comment by LucianCrainic** [comment]
Hi @Alan-MQ would you make this open? Looks like you got to an interesting point and i would love to take a look and see if i can help in anyway since i am really looking for this feature in VScode and it's the reason why i am still not switching from nvim to vscode.

**Comment by Alan-MQ** [comment]
> Hi [@Alan-MQ](https://github.com/Alan-MQ) would you make this open? Looks like you got to an interesting point and i would love to take a look and see if i can help in anyway since i am really looking for this feature in VScode and it's the reason why i am still not switching from nvim to vscode.

yeah , sure I've updated that comment, check it out, and feel free to let me know if I missed something~~

**Comment by BorisKaravasilev** [comment]
It would indeed be very useful to automatically show the preview of a searched file.

Are there any chances of this being prioritized any time soon @TylerLeonhardt? :)

**Comment by HolbyFPV** [comment]
https://github.com/jpcrs/binocular

Might be a good workaround if you can get it to work.  I wasn't having any luck on Windows, despite having all the dependencies.

**Comment by haxybaxy** [comment]
Hey guys, I just made this extension that allows you to preview the files while fuzzy searching through them, as well as fuzzy search and preview for most recent files opened: https://marketplace.visualstudio.com/items?itemName=zaidalsaheb.search-preview

This is my first time doing something like this, any feedback would be appreciated

**Comment by haxybaxy** [comment]
> Hey guys, I just made this extension that allows you to preview the files while fuzzy searching through them, as well as fuzzy search and preview for most recent files opened: [marketplace.visualstudio.com/items?itemName=zaidalsaheb.search-preview](https://marketplace.visualstudio.com/items?itemName=zaidalsaheb.search-preview)
> 
> This is my first time doing something like this, any feedback would be appreciated

If any of you are still looking for a fix for this, I've since updated this extension a lot and made it a ton better:
- Uses fzf under the hood so the search is really fast
- You can browse your most recent files like the oldfiles search in telescope
- Oldfiles search persists between sessions
- Works with all file types

I'm also actively maintaining it so if you have any feature requests or so you can leave them here:
https://github.com/haxybaxy/search-preview
